optimade/filtertransformers/mongo.py

```diff
@@ -81,20 +77,58 @@ def expression_phrase(self, arg):
         # expression_phrase: [ NOT ] ( comparison | "(" expression ")" )
         return self._recursive_expression_phrase(arg)
 
-    def property_first_comparison(self, arg):
+    @v_args(inline=True)
+    def property_first_comparison(self, quantity, query):
         # property_first_comparison: property ( value_op_rhs | known_op_rhs | fuzzy_string_op_rhs | set_op_rhs |
         # set_zip_op_rhs | length_op_rhs )
 
         # Awkwardly, MongoDB will match null fields in $ne filters,
         # so we need to add a check for null equality in evey $ne query.
-        if "$ne" in arg[1]:
-            return {"$and": [{arg[0]: arg[1]}, {arg[0]: {"$ne": None}}]}
-
-        return {arg[0]: arg[1]}
+        if "$ne" in query:
+            return {"$and": [{quantity: query}, {quantity: {"$ne": None}}]}
+
+        # Check if a $size query is being made (indicating a length_op_rhs filter); if so, check for
+        # a defined length alias to replace the $size call with the corresponding filter on the
+        # length quantity then carefully merge the two queries.
+        #
+        # e.g. `("elements", {"$size": 2, "$all": ["Ag", "Au"]})` should become
+        # `{"elements": {"$all": ["Ag", "Au"]}, "nelements": 2}` if the `elements` -> `nelements`
+        # length alias is defined.
+        if "$size" in query:
+            if (
+                getattr(self.backend_mapping.get(quantity), "length_quantity", None)
+                is not None
+            ):
+                size_query = {
+                    self.backend_mapping[
+                        quantity
+                    ].length_quantity.backend_field: query.pop("$size")
+                }
+
+                final_query = {}
+                if query:
+                    final_query = {quantity: query}
+                for q in size_query:
+                    if q in query:
+                        query[q].update(size_query[q])
+                    else:
+                        final_query[q] = size_query[q]
+
+                return final_query
+
+            # Another workaround: in the case that there is no length alias, and the field
+            # itself does not exist in the database, then `{"$size": 1}` matches all documents,
+            # so we must add another existence check
```

Comment:
Would it not be better to check whether the required size is 1? Otherwise, the exists check would be useless. 
I don't know how mongoDB will implement this query, but I can imagine that an extra exists check would take extra time.

Reply:
I considered this when initially writing it; we could hard-code a check for `{"$size": 1}` explicitly, but there are other cases like `{"$size": {"$lt": 2}}` etc that would also be problematic and I couldn't convince myself that I could exhaustively hard code them all. 

I would rather focus on robustness over speed in these transformers, especially for edge cases such as this, but I take your poitn.

NB: `{"field": {"$size": {"$lt": 2}}}` is not even a valid MongoDB query (you cannot apply operators to `$size`) but our filtertransformer uses this as an intermediate representation and turns it into the more convoluted query:

```json
{
    "$and": [
        {"field.1": {"$exists": True}},
        {"field.2": {"$exists": False}}
    ]
}
```

Reply:
Ok, I got a little carried away here, but I think this should cover all cases.
```
    python_operator_map = {"$lt":"<", "$lte":"<=", "$gt": ">", "$gte":">=", "$ne":"!=", "$eq":"==", "$in": " in ", "$nin" : " not in ", }
    sizearg = query.get("$size")
    if type(sizearg) == int : 
        evalstr = "1==" + str(sizearg)
    else: # case sizearg is dictionary
        evalstr = "1"+ python_operator_map.[str(list(sizearg.keys())[0])] + str(list(sizearg.values())[0])
    if (eval(evalstr)):
       ....
```

Could you also still address the second part of my comment?

Reply:
> Is the "$exits" operator the right operator to use here? It will also return documents where the property key is present but has the value null. I am not sure whether this is allowed with the OPTIMADE standard.

I don't think the behavior of LENGTH is well-defined on scalar or null fields in the specification, so we should make a sensible decision here. I agree just using `"$exists": True` may be problematic in the null but not missing value case, but I don't think I would want to implement the `eval` based solution you posted above.

I will add an additional `{"$ne": null}` as we did for the null value case (will comment here when I've pushed it).

Again, I just want to stress that this is an edge case, as any array field served by an implementation should be using `LENGTH_ALIASES`, and this is the case by default for all list fields specified by OPTIMADE.